src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs

```diff
@@ -370,18 +371,15 @@ private void CompleteStream(bool errored)
                     Log.RequestBodyNotEntirelyRead(ConnectionIdFeature, TraceIdentifier);
                 }
 
-                var (oldState, newState) = ApplyCompletionFlag(StreamCompletionFlags.Aborted);
+                var (oldState, newState) = ApplyCompletionFlag(StreamCompletionFlags.AbortedRead);
                 if (oldState != newState)
                 {
                     // https://quicwg.org/base-drafts/draft-ietf-quic-http.html#section-4.1-15
                     // When the server does not need to receive the remainder of the request, it MAY abort reading
                     // the request stream, send a complete response, and cleanly close the sending part of the stream.
                     // The error code H3_NO_ERROR SHOULD be used when requesting that the client stop sending on the
                     // request stream.
-
-                    // TODO(JamesNK): Abort the read half of the stream with H3_NO_ERROR
-                    // https://github.com/dotnet/aspnetcore/issues/33575
-
+                    _streamAbortFeature.AbortRead((long)Http3ErrorCode.NoError, new ConnectionAbortedException("The application completed without reading the entire request body."));
```

Comment:
If the application completed, it's done writing too, right? Can we just fully close the stream?

Or does this happen *only* when the body PipeReader was completed? If that's the case, maybe the message should say "The application stopped reading the request body." or something.

Reply:
> If the application completed, it's done writing too, right? Can we just fully close the stream?

The HTTP/3 spec says a server should abort reading if the server has completed processing.

> Or does this happen only when the body PipeReader was completed? If that's the case, maybe the message should say "The application stopped reading the request body." or something.

I have reused the logging message that the HTTP layer reports when a request is completed without reading the request body. It accurately describes why we're aborting reading.